### PR TITLE
[feature-metrics] Add global package that handles the metrics

### LINKED ISSUE #49: Feature: add global package for metrics
Add a global module that would be used to submit metrics to Prometheus.

### PR DESCRIPTION
**Pull request by ssnd**
closes #49

### CODE REVIEW

File: plugin/action/mask/mask.go [hunk 1]
```diff
@@ -148,26 +145,18 @@ func (p *Plugin) Start(config pipeline.AnyConfig, params *pipeline.ActionPluginP
 	p.valueNodes = make([]*insaneJSON.Node, 0)
 	p.logger = params.Logger
 	p.config.Masks = compileMasks(p.config.Masks, p.logger)
-	if p.config.MetricSubsystemName != nil {
```
Comment (ansakharov) [inline plugin/action/mask/mask.go:151]: Metric registration is required only if `MetricSubsystemName` is declared in config. This `if` can't be removed.

File: plugin/action/mask/mask.go [hunk 2]
```diff
@@ -255,8 +244,9 @@ func (p *Plugin) Do(event *pipeline.Event) pipeline.ActionResult {
 		}
 		v.MutateToString(string(p.maskBuf))
 	}
-	if p.logMaskAppeared && maskApplied {
```
Comment (ansakharov) [inline plugin/action/mask/mask.go:258]: Same as above: `p.logMaskAppeared` describes does mask appearance needs to be sent by Prometheus.

File: stats/stats.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+type stats struct {
+	allMetrics map[key]AnyMetric
+}
+
+var statsGlobal *stats
+
+func InitStats() {
+	statsGlobal = &stats{}
+	statsGlobal.allMetrics = make(map[key]AnyMetric)
```
Comment (ansakharov) [inline stats/stats.go:30]: Can do it shorter:
```go
statsGlobal = &stats{
    allMetrics: make(map[key]AnyMetric)
}
```

File: stats/stats.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+		Name:      metricDesc.Name,
+		Help:      metricDesc.Help,
+	})
+
+	keyInternal := key{pipeline.PromNamespace, metricDesc.Subsystem, metricDesc.Name}
+	statsGlobal.registerMetric(keyInternal, maskPromCounter)
+}
+
+func GetCounter(subsystem, metricName string) prom.Counter {
+	return statsGlobal.allMetrics[getKey(subsystem, metricName)].(prom.Counter)
```
Comment (anuriq) [inline stats/stats.go:47]: panic won't happen in any case?

File: Dockerfile [hunk 5]
```diff
@@ -7,4 +7,4 @@ WORKDIR /file.d
 
 COPY ./file.d .
 
-CMD ./file.d
+ENTRYPOINT [ "./file.d" ]
```
[the remark this answers is not in the document]
Comment (ssnd) [reply Dockerfile:10]: I changed this to be able to pass command line arguments in docker-compose without changing the executable file

[the remark this answers is not in the document]
Comment (anuriq) [reply Dockerfile]: i don't think this is needed. CMD can be overrided, and this is more expected than ENTRYPOINT

File: stats/stats.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+		Help:      metricDesc.Help,
+	})
+
+	keyInternal := key{pipeline.PromNamespace, metricDesc.Subsystem, metricDesc.Name}
+	registerMetric(keyInternal, maskPromCounter)
+}
+
+func GetCounter(subsystem, metricName string) prom.Counter {
+	if statsGlobal == nil {
+		logger.Panicf("stats package uninitialized")
```
Comment (ansakharov) [inline stats/stats.go:56]: Is it responsibility of GetCounter? Probably better return error and resolve on higher level.

Comment (vano144) [reply stats/stats.go:58]: Another option is initialization in init function in stats package.

File: stats/stats.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+	}
+
+	if val, ok := statsGlobal.allMetrics[getKey(subsystem, metricName)]; ok {
+		return val
+	}
+
+	return statsGlobal.allMetrics[getKey(subsystemName, unknownCounter)]
```
Comment (vano144) [inline stats/stats.go:65]: It can be calculated once

File: stats/stats.go [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+type key struct {
+	namespace string
+	subsystem string
+	name      string
+}
+
+type stats struct {
+	allMetrics map[key]prom.Counter //
```
Comment (vano144) [inline stats/stats.go:22]: // ?

File: plugin/output/elasticsearch/elasticsearch.go [hunk 9]
```diff
@@ -180,30 +205,39 @@ func (p *Plugin) out(workerData *pipeline.WorkerData, batch *pipeline.Batch) {
...
 		root, err := insaneJSON.DecodeBytes(respContent)
 		if err != nil {
+			stats.GetCounter(subsystemName, sendErrorCounter).Inc()
 			p.logger.Errorf("wrong response from %s, will try other endpoint: %s", endpoint, err.Error())
 			insaneJSON.Release(root)
 			continue
 		}
 
 		if root.Dig("errors").AsBool() {
+			errors := 0.0
```
Comment (vano144) [inline plugin/output/elasticsearch/elasticsearch.go:228]: Why it is float ?

File: plugin/output/kafka/kafka.go [hunk 10]
```diff
@@ -159,8 +175,9 @@ func (p *Plugin) out(workerData *pipeline.WorkerData, batch *pipeline.Batch) {
 		for _, e := range errs {
 			p.logger.Errorf("can't write batch: %s", e.Err.Error())
 		}
+		stats.GetCounter(subsystemName, sendErrorCounter).Add(float64(len(errs)))
```
Comment (vano144) [inline plugin/output/kafka/kafka.go:178]: Why did you decide not to do a similar check here?
https://github.com/ozontech/file.d/pull/50/files#diff-ae9e68585530e0cefac2588a53be8e80e41ff31d9f37ae0cbf67bc1e971c45c5R237

File: stats/stats.go (hunk 6 above)
Comment (ssnd) [reply stats/stats.go:58]: @anuriq @vano144 should we leave this panic here or remove it?

**Pull request merged**